### DISCUSSION TITLE
Handling Missing Directories When Generating Output

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#205] [General]
I found an issue with Gen-JSON where it fails to generate JSON files if the output directory path is missing one or more directories. I've captured the issue in more detail in #204. 

```
Gen-JSON fails to write JSON to file if the output file's parent directory doesn't exist. We should handle this more 
gracefully by creating the missing directories, interactively asking the user if they would like us to generate 
the directories, or failing with an error message that the directory is missing and needs to be created.
```

This issue exposes a larger issue/question for our users and their experience: _**how should we handle missing directories in user-specified output paths?**_ 

The three solutions I can think of are:
1. Automatically creating the missing directories for the user, they don't notice the issue at all.
2. Interactively ask the user if they would like us to generate the directories paired with an option to automatically generate or fail on missing directories via optional arguments `[--generate-missing, --fail-on-missing]`.
3. Fail with an error message that the directory is missing and needs to be created.

My personal preference is for option 2, especially if we pair it with user configuration in the future. The user would be able to configure their personal preference to always generate missing directories, always fail on missing directories, or interactively ask them for permission to generate the missing directories. The optional arguments also provide flexibility to opt-out of interactivity and local configuration for use with automated processes and pipelines.

My second choice is option 1, I don't think we need to inconvenience the user like with option 3. The user declared the directory when they ran the command, that should be sufficient trust to create a new directory if one doesn't exist.

I'd like to know your thoughts!

**Comment by cvchaparro** [comment]
I think that in every other plugin that writes output to directories, we follow option 1 so I think it makes sense to do that. That said just because we already do that doesn't mean we should continue to do that. In fact, I like the idea of option 2 (as an avid Emacs user, I always prefer getting to configure all the things exactly how I want). 

However, given our target user base and keeping in mind consistency with the rest of the plugins, I'd recommend we do option 1 and option 2. That is, default to auto-creating directories but create configuration options that allow users to turn it off, should they wish to.

**Reply by jondavid-black** [reply]
Maybe that configuration option would be something we put in the workspace configuration we discussed previously.  Is that what you were thinking?  I'd hate to have add a flag to all the plugins.

**Reply by cvchaparro** [reply]
Yeah, I was thinking we'd end up reading the configuration options from a file, at some point. As to whether it's a custom configuration file that can work with any "front-end" or a VSCode workspace-specific type of thing, I haven't really though of (although I think we'd have more flexibility to use the tool in workflows we haven't necessarily thought of if we just have a custom file).

That said, I think it may make sense to have flags people can pass to override their default configurations, if they want to. This would mean that, if they normally don't want any directories created, their configuration file would indicate that but they have the option to override it for a certain invocation, if needed, without having to temporarily change their configuration file.

It would definitely be a pain to have to pass a flag on each call to the tool for every configuration option, though, and I certainly wouldn't want to have to do that.

**Reply by Coffee2Bits** [reply]
I like the idea of configuration passed as CLI arguments being the top-level source of truth over personal/workspace configuration. We could have a tiered level with CLI args at the top, file-based config, then developer-provided defaults.

If we expose every file-based configuration option as an optional argument in the root argument parser then we should be able to declare those options once, and they should be compatible with all of the plugin commands assuming the plugins take advantage of those configuration values that we create and expose via an API.

**Comment by Coffee2Bits** [comment]
Given Cameron's feedback, it sounds like for now the standard plugin behavior should be to automatically create missing directories when generating output, and that we should begin planning the technical implementation of the workspaces proposal and/or some other file-based configuration option.
